hbase-server/src/main/java/org/apache/hadoop/hbase/master/webapp/MetaBrowser.java

```diff
@@ -0,0 +1,360 @@
+      .map(req -> req.getParameter(SCAN_LIMIT_PARAM))
+      .filter(StringUtils::isNotBlank);
+    if (!requestValueStr.isPresent()) { return null; }
+
+    final Integer requestValue = requestValueStr
+      .flatMap(MetaBrowser::tryParseInt)
+      .orElse(null);
+    if (requestValue == null) {
+      errorMessages.add(buildScanLimitMalformedErrorMessage(requestValueStr.get()));
+      return null;
+    }
+    if (requestValue <= 0) {
+      errorMessages.add(buildScanLimitLTEZero(requestValue));
+      return SCAN_LIMIT_DEFAULT;
+    }
+
+    final int truncatedValue = Math.min(requestValue, SCAN_LIMIT_MAX);
+    if (requestValue != truncatedValue) {
+      errorMessages.add(buildScanLimitExceededErrorMessage(requestValue));
+    }
+    return truncatedValue;
+  }
+
+  private RegionState.State resolveScanRegionState(final HttpServletRequest request) {
+    final Optional<String> requestValueStr = Optional.ofNullable(request)
+      .map(req -> req.getParameter(SCAN_REGION_STATE_PARAM))
+      .filter(StringUtils::isNotBlank)
+      .map(MetaBrowser::urlDecode);
+    if (!requestValueStr.isPresent()) { return null; }
+
+    final RegionState.State requestValue = requestValueStr
+      .flatMap(val -> tryValueOf(RegionState.State.class, val))
+      .orElse(null);
+    if (requestValue == null) {
+      errorMessages.add(buildScanRegionStateMalformedErrorMessage(requestValueStr.get()));
+      return null;
+    }
+    return requestValue;
+  }
+
+  private static byte[] resolveScanStart(final HttpServletRequest request) {
+    return Optional.ofNullable(request)
+      .map(req -> req.getParameter(SCAN_START_PARAM))
+      .filter(StringUtils::isNotBlank)
+      .map(MetaBrowser::urlDecode)
+      .map(Bytes::toBytesBinary)
+      .orElse(null);
+  }
+
+  private static TableName resolveScanTable(final HttpServletRequest request) {
+    return Optional.ofNullable(request)
+      .map(req -> req.getParameter(SCAN_TABLE_PARAM))
+      .filter(StringUtils::isNotBlank)
+      .map(MetaBrowser::urlDecode)
+      .map(TableName::valueOf)
+      .orElse(null);
+  }
+
+  private static Filter buildTableFilter(final TableName tableName) {
+    return new PrefixFilter(tableName.toBytes());
+  }
+
+  private static Filter buildScanRegionStateFilter(final RegionState.State state) {
+    return new SingleColumnValueFilter(
+      HConstants.CATALOG_FAMILY,
+      HConstants.TABLE_STATE_QUALIFIER,
+      CompareOperator.EQUAL,
+      // use the same serialization strategy as found in MetaTableAccessor#addRegionStateToPut
+      Bytes.toBytes(state.name()));
+  }
+
+  private Optional<Filter> buildScanFilter() {
+    if (scanTable == null && scanRegionState == null) {
+      return Optional.empty();
+    }
+
+    final List<Filter> filters = new ArrayList<>(2);
+    Optional.ofNullable(scanTable)
+      .map(MetaBrowser::buildTableFilter)
+      .ifPresent(filters::add);
+    Optional.ofNullable(scanRegionState)
```

Comment:
haha , functional style is not as readable :D (my personal opinion)

Reply:
Yeah, it's probably a bit overboard here. I started a back port to branch-1 and found there's only a couple places where Optional makes things easier.